library/pkparse.c

```diff
@@ -1194,12 +1197,24 @@ int mbedtls_pk_parse_key( mbedtls_pk_context *pk,
      * error
      */
 #if defined(MBEDTLS_PKCS12_C) || defined(MBEDTLS_PKCS5_C)
-    if( ( ret = pk_parse_key_pkcs8_encrypted_der( pk, key, keylen,
-                                                  pwd, pwdlen ) ) == 0 )
     {
-        return( 0 );
+        unsigned char *key_copy;
+
+        if( ( key_copy = mbedtls_calloc( 1, keylen ) ) == NULL )
```

Comment:
I see that there is no other choice here because `key` might be used again later on, but it seems a little pointless to be able to do in-place decryption at `pk_parse_key_pkcs8_encrypted_der()` and not being able to take advantage of it in this function.

This is just a suggestion, not a change request, but in this case I have a slight preference to simply move the buffer allocation inside `pk_parse_key_pkcs8_encrypted_der()` and drop the in-place operations. I could easily be convinced otherwise though...

Reply:
@andresag01 Looking at this code-path only, there's no difference between allocating here or within `pk_parse_key_pkcs8_encrypted_der`. However, at the other invocation of `pk_parse_key_pkcs8_encrypted_der` we already own the buffer passed to it, so another allocation in `pk_parse_key_pkcs8_encrypted_der` would be an unnecessary duplication.

Reply:
@hanno-arm: It was not a criticism, just an observation. This reduces the number of copies from 3 down to 2 at the most.

Reply:
@andresag01 I didn't take it as such, I only tried to explain my motivation for the change -
you're right in that the change is indeed only beneficial for one of the two occasions where `pk_parse_key_pkcs8_encrypted_der` is called, while keeping the memory usage in the other case as it was before.